src/resources/hash.cpp

```diff
@@ -0,0 +1,163 @@
+#include "resources/hash.hpp"
+#include "fileutils.hpp"
+#include "sha256.h"
+#include <memory>
+#include <optional>
+
+using namespace buddy::resources;
+
+class HashContext {
+private:
+    mbedtls_sha256_context sha;
+    uint32_t counter;
+
+public:
+    HashContext()
+        : sha()
+        , counter(0) {
+        mbedtls_sha256_init(&sha);
+        mbedtls_sha256_starts_ret(&sha, 0);
+    }
+
+    void append_mark() {
+        mbedtls_sha256_update_ret(&sha, (unsigned char *)&counter, sizeof(counter));
+        counter += 1;
+    }
+
+    void append_data(const uint8_t *data, size_t length) {
+        mbedtls_sha256_update_ret(&sha, data, length);
+    }
+
+    void finish(Hash &output) {
+        mbedtls_sha256_finish_ret(&sha, &output[0]);
+    }
+
+    ~HashContext() {
+        mbedtls_sha256_free(&sha);
+    }
+};
+
+static bool update_hash_with_file(const char *root, Path &path, HashContext &hash_ctx) {
+    // mark
+    hash_ctx.append_mark();
+
+    // filepath
+    size_t root_length = strlen(root);
+    const char *full_path = path.get();
+    const char *filepath = full_path + root_length;
+    hash_ctx.append_data((uint8_t *)filepath, strlen(filepath));
+
+    // mark
+    hash_ctx.append_mark();
+
+    // filedata
+    std::unique_ptr<FILE, FILEDeleter> source(fopen(path.get(), "rb"));
+    if (source.get() == nullptr) {
+        log(LOG_SEVERITY_ERROR, "Failed to open file %s", source.get());
+        return false;
+    }
+    uint8_t buffer[128];
+    bool success;
+    while (true) {
+        int read = fread(buffer, 1, sizeof(buffer), source.get());
+        if (read > 0) {
+            hash_ctx.append_data(buffer, read);
+        }
+        if (ferror(source.get())) {
+            success = false;
+            break;
+        }
+        if (feof(source.get())) {
+            success = true;
+            break;
+        }
+    }
+
+    // mark
+    hash_ctx.append_mark();
+
+    return success;
+}
+
+static bool update_hash_with_directory(const char *root, Path &path, HashContext &hash_ctx) {
+    std::optional<long> last_dir_location;
+
+    // mark
+    hash_ctx.append_mark();
+
+    // filepath
+    size_t root_length = strlen(root);
+    const char *full_path = path.get();
+    const char *filepath = full_path + root_length;
+    if (strlen(filepath) == 0) {
+        filepath = "/";
```

Comment:
Are you sure this pointer to a temporary variable `"/"` will be valid beyond the block where it is used? In reality it probably will, but it is a bad practice to rely on it.
Suggestion:
```
static const char rootS[] = "/";
...
filepath=rootS;
```

Reply:
Yes, the C++ standard says string literals have type of `const char[]` and have static storage duration.

Reply:
Oh, I didn't know this part of C++ standard. I must have run into some compiler issue back in the old days of "wooden computers" when this was not working thus being afraid of this construct ever since.